frontend/src/main/java/com/google/tripmeout/frontend/storage/PersistentPlaceVisitStorage.java

```diff
@@ -0,0 +1,140 @@
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.KeyFactory.Builder;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Transaction;
+import com.google.appengine.repackaged.com.google.datastore.v1.Datastore;
+import com.google.inject.Inject;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.error.PlaceVisitAlreadyExistsException;
+import com.google.tripmeout.frontend.error.PlaceVisitNotFoundException;
+import com.google.tripmeout.frontend.error.TripMeOutException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+class PersistentPlaceVisitStorage implements PlaceVisitStorage {
+  private static final String TRIP_ID_PROPERTY_NAME = "tripId";
+  private static final String USER_MARK_PROPERTY_NAME = "userMark";
+  private static final String ID_PROPERTY_NAME = "id";
+  private static final String PLACE_NAME_PROPERTY_NAME = "placeName";
+  private static final String PLACE_API_ID_PROPERTY_NAME = "placeApiId";
+
+  private final DatastoreService datastore;
+
+  @Inject
+  PersistentPlaceVisitStorage(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  public void addPlaceVisit(PlaceVisitModel placeVisit) throws PlaceVisitAlreadyExistsException {
+    Key placeKey = KeyFactory.stringToKey(placeVisit.id());
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      datastore.get(placeKey);
+      transaction.rollback();
+      throw new PlaceVisitAlreadyExistsException("Place Already Exists");
+    } catch (EntityNotFoundException e) {
+    }
+    try {
+      Entity placeEntity = new Entity(placeKey);
+      placeEntity.setProperty(ID_PROPERTY_NAME, placeVisit.id());
+      placeEntity.setProperty(TRIP_ID_PROPERTY_NAME, placeVisit.tripId());
+      placeEntity.setProperty(USER_MARK_PROPERTY_NAME, placeVisit.userMark().toString());
+      placeEntity.setProperty(PLACE_NAME_PROPERTY_NAME, placeVisit.placeName());
+      placeEntity.setProperty(PLACE_API_ID_PROPERTY_NAME, placeVisit.placesApiPlaceId());
+      datastore.put(placeEntity);
+    } catch (Exception e) {
+      transaction.rollback();
+      throw e;
+    }
+    transaction.commit();
+  }
+
+  public void removePlaceVisit(String tripId, String placeVisitId)
+      throws PlaceVisitNotFoundException {
+    Key placeKey = KeyFactory.stringToKey(placeVisitId);
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      datastore.get(placeKey);
+    } catch (EntityNotFoundException e) {
+      transaction.rollback();
+      throw new PlaceVisitNotFoundException("Place with id: " + placeVisitId + " doesn't exist");
+    }
+    datastore.delete(placeKey);
```

Comment:
Just to be safe, you should probably wrap this with a `try`/`finally`. I'm not sure how/if transactions are automatically aborted.